### DISCUSSION TITLE
I don't have worker logs and Queue/Email not working.

### DISCUSSION THREAD

**Discussion by mariuszmalek** [brefphp/bref#948] [Support] [answered]
Hi, five days I fight with configuration Bref. Queue not working and because of that, I can't send an email to a user after registering.
When I try to figurate what's going on then I realized don't have any logs file in the worker.

Can you help me?

**Comment by t-richard** [comment, marked as answer]
Hi @mariuszmalek !

Basically you should have :
- a SQS queue created in your account
- an IAM policy to allow your Lambda function to send messages to that queue
- a Lambda handler (worker) configured on that queue to consume messages

To troubleshoot the problem you could also check Cloudwatch metrics to see if
1. messages are sent to the queue (number of messages in the queue), if you don't see any then your main lambda is not sending messages to the queue or it is unable to do so
2. your Lambda worker is called (number of invocations), if they are none but you see messages in the queue, then it may mean that your worker is not configured to handle the messages from that queue

Hope that helps, if not then please provide us with more details like samples of your configuration, framework used if any, how is your SQS queue created, how is your worker configured, etc :slightly_smiling_face:

**Reply by mariuszmalek** [reply]
I am are really grateful for your answer.

Sure I create SQS in AWS and do everything that is in the documentation + I try configuration from the youtube tutorial.
The problem is after register (when the software send a welcome email), then I have an error and the software doesn't go to next, but stay in one place (register form). I don't see any new logs in the worker function (cloudwatch).

Below I upload a screenshot from the source code.

![image](https://user-images.githubusercontent.com/26696860/120990732-25928500-c781-11eb-80a4-2d77bca1df0a.png)
![image](https://user-images.githubusercontent.com/26696860/120990767-30e5b080-c781-11eb-87b7-8b7874b6334e.png)
![image](https://user-images.githubusercontent.com/26696860/120990813-422ebd00-c781-11eb-82ed-1e63b3ac9b79.png)
![image](https://user-images.githubusercontent.com/26696860/120990846-4ce95200-c781-11eb-937c-5451c330e285.png)
![image](https://user-images.githubusercontent.com/26696860/120990897-5a064100-c781-11eb-88b2-38f464a0e1d4.png)

**Reply by mariuszmalek** [reply]
![image](https://user-images.githubusercontent.com/26696860/120991434-eca6e000-c781-11eb-8ece-ffaa52d7851e.png)

**Reply by t-richard** [reply]
I'm not super familiar with Laravel but it seems to me that in what I believe is your `index.php` file, you are using `getenv('SQS_PREFIX').'/'.getenv('SQS_QUEUE')`

In the way you defined your `SQS_QUEUE` env variable in `serverless.yml` it seems to me that the URL would look something like `/https://sqs.eu-central-1.amazonaws.com/....` (starting with a `/`).

This would certainly cause Laravel to crash because it's an invalid URL. 

Then it is also a normal thing not to see logs in you worker's Cloudwatch log group as it is never invoked by AWS because there is no message in the queue.

But I guess you should have logs in your web function's Cloudwatch log group because it is where the error occurs.

Let me know if that helped you and feel free to ask again if you still have troubles or if it isn't clear enough :slightly_smiling_face:

**Reply by mariuszmalek** [reply]
I changed this as you said (still I have an error when I try to register). I think, still, Laravel can't send anything to SQS. When I run the test job from the Lambda function the first time I see any logs.

![image](https://user-images.githubusercontent.com/26696860/121002029-309ee280-c78c-11eb-930e-68a70da60400.png)

**Reply by t-richard** [reply]
This is pure Laravel here, I might not be able to help or say complete non-sense but I'll try my best :sweat_smile: .

This looks like worker function logs, not web function logs. In the Stack trace, you can see it fails at line 96 in Job from laravel Queue which means the job payload is empty https://github.com/laravel/framework/blob/8.x/src/Illuminate/Queue/Jobs/Job.php#L96 (trying to access the `job` key on `$payload` which is apparently null).

Which probably means that the laravel bridge is getting an empty body here https://github.com/brefphp/laravel-bridge/blob/master/src/Queue/LaravelSqsHandler.php#L62 and thus that your SQS message is empty.

I also see "This handler expected to be invoked with a SQS event" in the logs, maybe expanding this trace would give you more insight on what's going on. If your handler is configured for SQS only, this should not happen IMHO.

**Reply by mariuszmalek** [reply]
Empty? I send by Lambda test to SQS. Sending by my application not working (I don't see anything in logs)

![image](https://user-images.githubusercontent.com/26696860/121017615-c3944880-c79d-11eb-8abb-ae903983b8a9.png)

**Reply by mariuszmalek** [reply]
![image](https://user-images.githubusercontent.com/26696860/121017757-eb83ac00-c79d-11eb-9080-9cce964dde36.png)

**Reply by t-richard** [reply]
I think there is a confusion here.

You are actually sending a test event to Lambda, not to SQS so the error is totally legitimate.

This particular Lambda function is expecting a SQS event that has a specific format, you can see an example of a SQS event here https://docs.aws.amazon.com/lambda/latest/dg/with-sqs.html

When you are sending a plain array as a Lambda event, the handler fails because it is not a valid SQS event. Hope that makes sense.

To test the Lambda <-> SQS integration, go in your SQS console, select the desired Queue and click on the `Send and receive messages` button then add whatever you need in the body. This should trigger the lambda worker.

note that doing so might also make your code crash because Laravel might be expecting a specific format and that I don't know.

**Reply by mnapoli** [reply]
> To test the Lambda <-> SQS integration, go in your SQS console, select the desired Queue and click on the Send and receive messages button then add whatever you need in the body. This should trigger the lambda worker.

Indeed this will also cause Laravel to crash, because the job will still not be valid.

The job needs to be triggered by Laravel.

**Reply by mariuszmalek** [reply]
Thanks for the answers! I now understand and Queue works well, but I still have a problem with sending e-mail. When I try to register in the software then I have a bug in one place an e-mail wasn't sending.

**Reply by t-richard** [reply]
@mariuszmalek we don't know your app so helping you with this few details is really hard.

Do you have an error ? Does the Lambda function timeout ? Do you have logs ? Is this done in the queue ? What are you using to send emails (which provider) ? etc
